modules/4337/contracts/experimental/verifiers/WebAuthnVerifier.sol

```diff
@@ -102,17 +104,47 @@ contract WebAuthnVerifier is IWebAuthnVerifier, P256Wrapper {
         bytes32 challenge,
         bytes calldata clientDataFields
     ) internal pure returns (bytes32 message) {
-        string memory encodedChallenge = Base64Url.encode(abi.encodePacked(challenge));
+
         /* solhint-disable quotes */
+        // AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA is placeholder for the encoded challenge
         bytes memory clientDataJson = abi.encodePacked(
             '{"type":"webauthn.get","challenge":"',
-            encodedChallenge,
+            "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
             '",',
             clientDataFields,
             "}"
         );
+
+        string memory table = ENCODING_TABLE;
+        
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // Skip first 32 bytes of the table containing the length
+            let tablePtr := add(table, 1)
+            // Skip first 36 bytes of the clientDataJson containing '{"type":"webauthn.get","challenge":"'
+            let resultPtr := add(clientDataJson, 68)
+
+            // Store group of 6 bits from the challenge to be encoded. Storing of 6 bits group can be removed but, kept here for readability.
+            let sixBitGroup
+
+            // Iterate over challenge in group of 6 bits, for each 6 bits lookup the ENCODING_TABLE, transform it and store it in the result
+            for {let i := 250} lt(i, 251)
+            {
+                i := sub(i, 6)
+            } {
+                sixBitGroup := and(shr(i, challenge), 0x3F)
+                mstore8(resultPtr, mload(add(tablePtr, sixBitGroup)))
+                resultPtr := add(resultPtr, 1)
+            }
+
+            // Load the remaining last 4 bits of challenge that are yet to be encoded and then shift left to add 2 bits at the end to make it a group of 6 bits.
+            sixBitGroup := shl(2, and(challenge, 0x0F))
```

Comment:
I wouldn't mix decimals and hexadecimal in assembly and would stick to one numbering system to ease processing and remove switching contexts